File: java/concept/answer/ConceptMapImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.driver.concept.answer;

import com.vaticle.typedb.driver.api.answer.ConceptMap;
import com.vaticle.typedb.driver.api.concept.Concept;
import com.vaticle.typedb.driver.common.NativeObject;
import com.vaticle.typedb.driver.common.NativeIterator;
import com.vaticle.typedb.driver.common.exception.TypeDBDriverException;
import com.vaticle.typedb.driver.concept.ConceptImpl;
import com.vaticle.typedb.common.collection.Pair;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.vaticle.typedb.driver.common.exception.ErrorMessage.Concept.MISSING_VARIABLE;
import static com.vaticle.typedb.driver.common.exception.ErrorMessage.Concept.NONEXISTENT_EXPLAINABLE_CONCEPT;
import static com.vaticle.typedb.driver.common.exception.ErrorMessage.Concept.NONEXISTENT_EXPLAINABLE_OWNERSHIP;
import static com.vaticle.typedb.driver.common.exception.ErrorMessage.Query.VARIABLE_DOES_NOT_EXIST;
import static com.vaticle.typedb.driver.jni.typedb_driver.concept_map_equals;
import static com.vaticle.typedb.driver.jni.typedb_driver.concept_map_get;
import static com.vaticle.typedb.driver.jni.typedb_driver.concept_map_get_explainables;
import static com.vaticle.typedb.driver.jni.typedb_driver.concept_map_get_values;
import static com.vaticle.typedb.driver.jni.typedb_driver.concept_map_get_variables;
import static com.vaticle.typedb.driver.jni.typedb_driver.concept_map_to_string;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainable_get_conjunction;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainable_get_id;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_equals;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_get_attribute;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_get_attributes_keys;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_get_ownership;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_get_ownerships_keys;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_get_relation;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_get_relations_keys;
import static com.vaticle.typedb.driver.jni.typedb_driver.explainables_to_string;

public class ConceptMapImpl extends NativeObject<com.vaticle.typedb.driver.jni.ConceptMap> implements ConceptMap {
    private int hash = 0;
    private Map<String, Concept> cachedMap = null;

    public ConceptMapImpl(com.vaticle.typedb.driver.jni.ConceptMap concept_map) {
        super(concept_map);
    }

    @Override
    public Stream<String> variables() {
        return new NativeIterator<>(concept_map_get_variables(nativeObject)).stream();
    }

    @Override
    public Stream<Concept> concepts() {
        return new NativeIterator<>(concept_map_get_values(nativeObject)).stream().map(ConceptImpl::of);
    }

    @Override
    public Map<String, Concept> map() {
        if (cachedMap == null) cachedMap = variables().collect(Collectors.toUnmodifiableMap(v -> v, this::get));
        return cachedMap;
    }

    @Override
    public Concept get(String variable) {
        if (variable == null || variable.isEmpty()) throw new TypeDBDriverException(MISSING_VARIABLE);
        com.vaticle.typedb.driver.jni.Concept concept = concept_map_get(nativeObject, variable);
        if (concept == null) throw new TypeDBDriverException(VARIABLE_DOES_NOT_EXIST, variable);
        return ConceptImpl.of(concept);
    }

    @Override
    public Explainables explainables() {
        return new ExplainablesImpl(concept_map_get_explainables(nativeObject));
    }

    @Override
    public String toString() {
        return concept_map_to_string(nativeObject);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        ConceptMapImpl that = (ConceptMapImpl) obj;
        return concept_map_equals(this.nativeObject, that.nativeObject);
    }

    @Override
    public int hashCode() {
        if (hash == 0) hash = computeHash();
        return hash;
    }

    private int computeHash() {
        return map().hashCode();
    }

    public static class ExplainablesImpl extends NativeObject<com.vaticle.typedb.driver.jni.Explainables> implements Explainables {
        private int hash = 0;

        ExplainablesImpl(com.vaticle.typedb.driver.jni.Explainables explainables) {
            super(explainables);
        }

        @Override
        public Explainable relation(String variable) {
            if (variable == null || variable.isEmpty()) throw new TypeDBDriverException(MISSING_VARIABLE);
            com.vaticle.typedb.driver.jni.Explainable explainable = explainables_get_relation(nativeObject, variable);
            if (explainable == null) throw new TypeDBDriverException(NONEXISTENT_EXPLAINABLE_CONCEPT, variable);
            return new ExplainableImpl(explainable);
        }

        @Override
        public Explainable attribute(String variable) {
            if (variable == null || variable.isEmpty()) throw new TypeDBDriverException(MISSING_VARIABLE);
            com.vaticle.typedb.driver.jni.Explainable explainable = explainables_get_attribute(nativeObject, variable);
            if (explainable == null) throw new TypeDBDriverException(NONEXISTENT_EXPLAINABLE_CONCEPT, variable);
            return new ExplainableImpl(explainable);
        }

        @Override
        public Explainable ownership(String owner, String attribute) {
            if (owner == null || owner.isEmpty()) throw new TypeDBDriverException(MISSING_VARIABLE);
            if (attribute == null || attribute.isEmpty()) throw new TypeDBDriverException(MISSING_VARIABLE);
            com.vaticle.typedb.driver.jni.Explainable explainable = explainables_get_ownership(nativeObject, owner, attribute);
            if (explainable == null) throw new TypeDBDriverException(NONEXISTENT_EXPLAINABLE_OWNERSHIP, owner, attribute);
            return new ExplainableImpl(explainable);
        }

        @Override
        public Stream<Pair<String, Explainable>> relations() {
            return new NativeIterator<>(explainables_get_relations_keys(nativeObject)).stream().map(k -> new Pair<>(k, relation(k)));
        }

        @Override
        public Stream<Pair<String, Explainable>> attributes() {
            return new NativeIterator<>(explainables_get_attributes_keys(nativeObject)).stream().map(k -> new Pair<>(k, attribute(k)));
        }

        @Override
        public Stream<Pair<Pair<String, String>, Explainable>> ownerships() {
            return new NativeIterator<>(explainables_get_ownerships_keys(nativeObject)).stream().map(pair -> {
                String owner = pair.get_0();
                String attribute = pair.get_1();
                return new Pair<>(new Pair<>(owner, attribute), ownership(owner, attribute));
            });
        }

        @Override
        public String toString() {
            return explainables_to_string(nativeObject);
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            ExplainablesImpl that = (ExplainablesImpl) obj;
            return explainables_equals(this.nativeObject, that.nativeObject);
        }

        @Override
        public int hashCode() {
            if (hash == 0) hash = computeHash();
            return hash;
        }

        private int computeHash() {
            return Objects.hash(
                    relations().collect(Collectors.toList()),
                    attributes().collect(Collectors.toList()),
                    ownerships().collect(Collectors.toList())
            );
        }
    }

    public static class ExplainableImpl extends NativeObject<com.vaticle.typedb.driver.jni.Explainable> implements Explainable {
        public ExplainableImpl(com.vaticle.typedb.driver.jni.Explainable explainable) {
            super(explainable);
        }

        @Override
        public String conjunction() {
            return explainable_get_conjunction(nativeObject);
        }

        @Override
        public long id() {
            return explainable_get_id(nativeObject);
        }

        @Override
        public String toString() {
            return "Explainable { id: " + id() + ", conjunction: " + conjunction() + " }";
        }


        @Override
        public boolean equals(final Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            final ExplainableImpl that = (ExplainableImpl) obj;
            return this.id() == that.id();
        }

        @Override
        public int hashCode() {
            return (int) id();
        }
    }
}

File: nodejs/concept/answer/ConceptMapImpl.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
    ConceptMap as ConceptMapProto,
    Explainable as ExplainableProto,
    Explainables as ExplainablesProto,
    ExplainablesOwned as ExplainablesOwnedProto
} from "typedb-protocol/proto/answer";
import {Concept as ConceptProto} from "typedb-protocol/proto/concept";
import {ConceptMap} from "../../api/answer/ConceptMap";
import {Concept} from "../../api/concept/Concept";
import {ErrorMessage} from "../../common/errors/ErrorMessage";
import {TypeDBDriverError} from "../../common/errors/TypeDBDriverError";
import {ResponseReader} from "../../common/rpc/ResponseReader";

export class ConceptMapImpl implements ConceptMap {
    private readonly _concepts: Map<string, Concept>;
    private readonly _explainables: ConceptMap.Explainables;

    constructor(concepts: Map<string, Concept>, explainables: ConceptMap.Explainables) {
        this._concepts = concepts;
        this._explainables = explainables;
    }

    variables(): IterableIterator<string> {
        return this._concepts.keys();
    }

    concepts(): IterableIterator<Concept> {
        return this._concepts.values();
    }

    map(): Map<string, Concept> {
        return this._concepts;
    }

    get(variable: string): Concept {
        return this._concepts.get(variable);
    }

    get explainables(): ConceptMap.Explainables {
        return this._explainables;
    }
}

/* eslint no-inner-declarations: "off" */
export namespace ConceptMapImpl {
    import NONEXISTENT_EXPLAINABLE_CONCEPT = ErrorMessage.Query.NONEXISTENT_EXPLAINABLE_CONCEPT;
    import NONEXISTENT_EXPLAINABLE_OWNERSHIP = ErrorMessage.Query.NONEXISTENT_EXPLAINABLE_OWNERSHIP;

    export function of(proto: ConceptMapProto): ConceptMap {
        const variableMap = new Map<string, Concept>();
        proto.map.forEach((proto: ConceptProto, resLabel: string) => variableMap.set(resLabel, ResponseReader.Concept.of(proto)));
        const explainables = proto.has_explainables ? ofExplainables(proto.explainables) : emptyExplainables();
        return new ConceptMapImpl(variableMap, explainables);
    }

    function ofExplainables(proto: ExplainablesProto): ConceptMap.Explainables {
        const relations = new Map<string, ConceptMap.Explainable>();
        proto.relations.forEach((explainable: ExplainableProto, variable: string) =>
            relations.set(variable, ofExplainable(explainable))
        );
        const attributes = new Map<string, ConceptMap.Explainable>();
        proto.attributes.forEach((explainable: ExplainableProto, variable: string) =>
            relations.set(variable, ofExplainable(explainable))
        );
        const ownerships = new Map<[string, string], ConceptMap.Explainable>();
        proto.ownerships.forEach((owned: ExplainablesOwnedProto, owner: string) =>
            owned.owned.forEach((explainable: ExplainableProto, attribute: string) => {
                ownerships.set([owner, attribute], ofExplainable(explainable))
            })
        );
        return new ExplainablesImpl(relations, attributes, ownerships)
    }

    function emptyExplainables() {
        return new ExplainablesImpl(
            new Map<string, ConceptMap.Explainable>(),
            new Map<string, ConceptMap.Explainable>(),
            new Map<[string, string], ConceptMap.Explainable>()
        );
    }

    function ofExplainable(proto: ExplainableProto): ConceptMap.Explainable {
        return new ExplainableImpl(proto.conjunction, proto.id);
    }

    export class ExplainablesImpl implements ConceptMap.Explainables {
        private readonly _relations: Map<string, ConceptMap.Explainable>;
        private readonly _attributes: Map<string, ConceptMap.Explainable>;
        private readonly _ownerships: Map<[string, string], ConceptMap.Explainable>;

        constructor(relations: Map<string, ConceptMap.Explainable>, attributes: Map<string, ConceptMap.Explainable>,
                    ownerships: Map<[string, string], ConceptMap.Explainable>) {
            this._relations = relations;
            this._attributes = attributes;
            this._ownerships = ownerships;
        }

        relation(variable: string): ConceptMap.Explainable {
            const explainable = this._relations.get(variable);
            if (!explainable) throw new TypeDBDriverError(NONEXISTENT_EXPLAINABLE_CONCEPT.message(variable));
            return explainable;
        }

        attribute(variable: string): ConceptMap.Explainable {
            const explainable = this._attributes.get(variable);
            if (!explainable) throw new TypeDBDriverError(NONEXISTENT_EXPLAINABLE_CONCEPT.message(variable));
            return explainable;
        }

        ownership(owner: string, attribute: string): ConceptMap.Explainable {
            for (const entry of this._ownerships) {
                if (entry[0][0] === owner && entry[0][1] === attribute) return entry[1];
            }
            throw new TypeDBDriverError(NONEXISTENT_EXPLAINABLE_OWNERSHIP.message(owner, attribute));
        }

        get relations(): Map<string, ConceptMap.Explainable> {
            return this._relations;
        }

        get attributes(): Map<string, ConceptMap.Explainable> {
            return this._attributes;
        }

        get ownerships(): Map<[string, string], ConceptMap.Explainable> {
            return this._ownerships;
        }
    }

    export class ExplainableImpl implements ConceptMap.Explainable {
        private readonly _conjunction: string;
        private readonly _id: number;

        constructor(conjunction: string, id: number) {
            this._conjunction = conjunction;
            this._id = id;
        }

        get conjunction(): string {
            return this._conjunction;
        }

        get id(): number {
            return this._id;
        }
    }
}
